text/0000-variadic-generics.md

```diff
@@ -0,0 +1,444 @@
+in `T` must satisfy the bound `Trait`.
+
+```rust
+fn example<...T: Clone>(...args: T);
+```
+
+Now that the types in `T` can be bound and the list of arguments may be unpacked
+it becomes possible to iterate over the list of arguments and call trait methods
+on each argument in the list.
+
+```rust
+fn example<...T: Clone>(...args: T) {
+    // All types in T implement clone, so we can use
+    // the clone method on the unpacked args
+    f(...args.clone());
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When working with variadics it is helpful to be able to:
+
+ - unpack the arguments
+ - recurse over the arguments
+ - fold over the arguments
+
+## Recursion
+[recursion]: #recursion
+
+In order to allow recursing over the list there needs to be a way to
+extract an argument (the first argument in most cases). The addition
+of the pattern `...arg` to represent a argument that "captures"
+multiple arguments can be employed to allow this. When used as an
+l-value pattern to an expression, the variable `...arg` in a tuple
+pattern will consume the remainder of elements in the variadic list
+of arguments. For example, a simple recursive function might look
+something like the following:
+
+```rust
+fn recurse<...T>(...args: T) {
+   let (head, ...tail) = args;
+   // Do something with `head`
+   recurse(...tail);
+}
+```
+
+### Unpacking
+[unpacking]: #unpacking
+
+**TODO(dlrobertson)**: Is user constructed `Tuple` unpacking valuable?
+Is this relevant to the RFC? Or should we only allow this for variadic
+generic tuples? After some thought [solution-3] seems to be the best
+option. So the discussion surrounding unpacking user defined tuples
+may be cut.
+
+Note that adding the ability to unpack the `Tuple` of variadic arguments
+may also impact user defined `Tuple`s. Due to the layout of `Tupple`
+types and the possible introduction of padding to propperly align types,
+the type of the value returned from unpacking a `Tupple` should be
+carefully considered.
+
+#### Solution 1
+[solution-1]: #solution-1
+
+When capturing the remaining elements of a `Tupple` with the `...`
+pattern, the result will capture the tail by reference.
+
+```rust
+let some_tuple = (42u8, 65535u16, 0u8);
+let (head, ...tail) = some_tuple; // head: u8, tail: (&u16, &u8)
+
+// If the extracted value of the tuple implements `Copy`
+// then the original `Tuple` has not been moved.
+do_something(some_tuple);
+```
+
+**Limitations**: Capturing the tail by reference going to cause
+us problems later? E.g., what do we do here:
+
+
+```rust
+let some_tuple = (42u8, 0u8, 65535u16, 0u8)
+let (_, ...t0) = some_tuple // t0: (&u8, &u16, &u8)
+let (_, ...t1) = t0 // What is t1? is it (&&u16, &&u8)?
+```
+
+This will increase the number of instructions need to unpack, as a
+new `Tuple` will need to be constructed.
+
+#### Solution 2
+[solution-2]: #solution-2
```

Comment:
This seems more viable and a less invasive change in the language overall in the type system.